### DISCUSSION TITLE
Can I control the map by foxglove-extension?

### DISCUSSION THREAD

**Discussion by WuMianzhi** [foxglove/community#1024] [General]
I want to customize a Foxglove extension that can modify the map's behavior, such as changing the zoom level or other interactions. However, it seems there aren't any related docs or APIs available. Is there a way to achieve this?

**Comment by jtbandes** [comment]
Extensions can't modify existing panels, only convert messages or install custom panels. Could you say more about what you are trying to accomplish?

**Reply by WuMianzhi** [reply]
Thanks for your reply. I'm trying to preload all GPS data ( jsut like map pannal ) when the extension is loaded and seek the nearest GPS data to my target location. However, when I set the preload property to true using `context.subscribe([{ topic: "/gps", preload: true }]);` and then use `renderState.allFrames`, it seems like I can't get all the frames. Is this a bug, or did I write the code incorrectly? Here is my code:
```jsx
import { Immutable, MessageEvent, PanelExtensionContext, Topic } from "@foxglove/extension";
// import { LocationFix } from "@foxglove/schemas";
import { useEffect, useLayoutEffect, useState } from "react";
import ReactDOM from "react-dom";

// type LocationFrame = MessageEvent<LocationFix>;

function ExamplePanel({ context }: { context: PanelExtensionContext }): JSX.Element {
  const [topics, setTopics] = useState<undefined | Immutable<Topic[]>>();
  const [messages, setMessages] = useState<undefined | Immutable<MessageEvent[]>>();

  const [renderDone, setRenderDone] = useState<(() => void) | undefined>();
  // const [allFrames, setAllFrames] = useState<LocationFrame[]>([]);
  const [frameCount, setFrameCount] = useState(0);
  const [startTime, setStartTime] = useState("0");
  const [endTime, setEndTime] = useState("0");

  useLayoutEffect(() => {
    context.onRender = (renderState, done) => {
      setStartTime(() => renderState.startTime?.sec + "." + renderState.startTime?.nsec);
      setEndTime(() => renderState.endTime?.sec + "." + renderState.endTime?.nsec);

      if (renderState.allFrames) {
        setFrameCount((count) => count + (renderState.allFrames?.length ?? 0));
      }

      setRenderDone(() => done);
      setTopics(renderState.topics);
      setMessages(renderState.currentFrame);
    };

    context.watch("topics");
    context.watch("currentFrame");
    context.subscribe([{ topic: "/gps", preload: true }]);

    return () => {
      context.unsubscribeAll();
    };
  }, [context]);

  // invoke the done callback once the render is complete
  useEffect(() => {
    renderDone?.();
  }, [renderDone]);

  const changeTime = () => {
    // Example: seek to 10 seconds
    if (context.seekPlayback) {
      context.seekPlayback(1532449727);
      console.log("play back to 6");
    }
  };

  return (
    <div style={{ padding: "1rem" }}>
      <h2>Welcome to your new extension panel!</h2>
      <div>
        <button onClick={changeTime}>Seek to 10s</button>
        <p>
          playedFrameCount: <span>{frameCount}</span>
        </p>
      </div>
      <p>startTime: {startTime}</p>
      <p>endTime: {endTime}</p>
      <p>
        Check the{" "}
        <a href="https://foxglove.dev/docs/studio/extensions/getting-started">documentation</a> for
        more details on building extension panels for Foxglove Studio.
      </p>
      <div style={{ display: "grid", gridTemplateColumns: "1fr 1fr", rowGap: "0.2rem" }}>
        <b style={{ borderBottom: "1px solid" }}>Topic</b>
        <b style={{ borderBottom: "1px solid" }}>Schema name</b>
        {(topics ?? []).map((topic) => (
          <>
            <div key={topic.name}>{topic.name}</div>
            <div key={topic.schemaName}>{topic.schemaName}</div>
          </>
        ))}
      </div>
      <div>{messages?.length}</div>
    </div>
  );
}

export function initExamplePanel(context: PanelExtensionContext): () => void {
  ReactDOM.render(<ExamplePanel context={context} />, context.panelElement);

  // Return a function to run when the panel is removed
  return () => {
    ReactDOM.unmountComponentAtNode(context.panelElement);
  };
}
```

and the framecount will be 0 even if you play the file, which means allframes length always be zero

**Reply by defunctzombie** [reply]
You are missing a `context.watch("allFrames");`. Similar to how you watch the topics and current frame you also need to tell the context you want allFrames.

**Reply by WuMianzhi** [reply]
I got it, thank you very much!!!, Is there any way to know that allframes has been loaded?

**Reply by defunctzombie** [reply]
No - unfortunately there is not yet a way to know it has been loaded.

**Reply by WuMianzhi** [reply]
one more question, if i use `context.watch("allFrames")` which means i dont need to `context.watch("currentFrame")` ?

**Reply by defunctzombie** [reply]
If you are only going to use `allFrames` and don't have any logic that deals with `currentFrame` then yea you don't need `currentFrame`. However - `allFrames` is not available for "live" connections (like websocket) so if you want your panel to work with live data and replay data then you still do need currentFrame.

**Reply by WuMianzhi** [reply]
Ok, thank you again! Maybe when I finish my current project, I could write a tutorial similar to [building-a-custom-react-panel-with-foxglove-studio-extensions](https://foxglove.dev/blog/building-a-custom-react-panel-with-foxglove-studio-extensions). If you'd like, you could add it to that series so it can help other developers.
